File: brdgme-go/greed/command.go
package greed

import (
	"errors"
	"fmt"
	"strings"

	"github.com/brdgme/brdgme/brdgme-go/brdgme"
	"github.com/brdgme/brdgme/brdgme-go/libdie"
)

type ScoreCommand struct {
	Dice []int
}

type RollCommand struct{}

type DoneCommand struct{}

func (g *Game) Command(
	player int,
	input string,
	players []string,
) (brdgme.CommandResponse, error) {
	parser := g.CommandParser(player)
	if parser == nil {
		return brdgme.CommandResponse{}, errors.New(
			"not expecting any commands at the moment",
		)
	}
	output, err := parser.Parse(input, players)
	if err != nil {
		return brdgme.CommandResponse{}, err
	}
	switch value := output.Value.(type) {
	case ScoreCommand:
		return g.ScoreCommand(player, value.Dice, output.Remaining)
	case RollCommand:
		return g.RollCommand(player, output.Remaining)
	case DoneCommand:
		return g.DoneCommand(player, output.Remaining)
	}
	return brdgme.CommandResponse{}, errors.New("inexhaustive command handler")
}

func (g *Game) CommandSpec(player int) *brdgme.Spec {
	parser := g.CommandParser(player)
	if parser != nil {
		spec := parser.ToSpec()
		return &spec
	}
	return nil
}

func (g *Game) ScoreCommand(player int, dice []int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Score(player, dice)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   true,
		Remaining: remaining,
	}, err
}

func (g *Game) RollCommand(player int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.PlayerRoll(player)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) DoneCommand(player int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Done(player)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) CommandParser(player int) brdgme.Parser {
	parsers := brdgme.OneOf{}
	if g.CanScore(player) {
		parsers = append(parsers, g.ScoreParser())
	}
	if g.CanRoll(player) {
		parsers = append(parsers, RollParser)
	}
	if g.CanDone(player) {
		parsers = append(parsers, DoneParser)
	}
	if len(parsers) == 0 {
		return nil
	}
	return parsers
}

func ScoreDiceParser(score Score) brdgme.Parser {
	diceStrs := make([]string, len(score.Dice))
	for i, d := range score.Dice {
		diceStrs[i] = DieNames[d]
	}
	name := strings.Join(diceStrs, "")
	return brdgme.Doc{
		Name: name,
		Desc: fmt.Sprintf("%d points", score.Value),
		Parser: brdgme.Map{
			Parser: brdgme.Token(name),
			Func: func(value interface{}) interface{} {
				return score.Dice
			},
		},
	}
}

func (g *Game) ScoreParser() brdgme.Parser {
	available := []brdgme.Parser{}
	for _, s := range Scores() {
		if ok, _ := libdie.DiceInDice(s.Dice, g.RemainingDice); ok {
			available = append(available, ScoreDiceParser(s))
		}
	}
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "score",
				Desc:   "score dice",
				Parser: brdgme.Token("score"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "dice",
					Desc:   "the dice to score",
					Parser: brdgme.OneOf(available),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return ScoreCommand{
				Dice: value.([]interface{})[1].([]Die),
			}
		},
	}
}

var RollParser = brdgme.Doc{
	Name: "roll",
	Desc: "roll the dice",
	Parser: brdgme.Map{
		Parser: brdgme.Token("roll"),
		Func: func(value interface{}) interface{} {
			return RollCommand{}
		},
	},
}

var DoneParser = brdgme.Doc{
	Name: "done",
	Desc: "finish your turn",
	Parser: brdgme.Map{
		Parser: brdgme.Token("done"),
		Func: func(value interface{}) interface{} {
			return DoneCommand{}
		},
	},
}

func DieParser() brdgme.Parser {
	values := make([]brdgme.EnumValue, len(DieFaces))
	for i, die := range DieFaces {
		values[i] = brdgme.EnumValue{
			Value: die,
			Name:  DieNames[die],
		}
	}
	return brdgme.Enum{
		Values: values,
	}
}

File: rust/game/greed-2/src/command.rs
use brdgme_game::command::parser::*;

use crate::Game;
use crate::{Die, Score};

#[derive(Debug, PartialEq, Clone)]
pub enum Command {
    Score { dice: Vec<Die> },
    Roll,
    Done,
}

impl Game {
    pub fn command_parser(&self, player: usize) -> Option<Box<dyn Parser<T = Command> + '_>> {
        if self.finished() {
            return None;
        }
        let mut parsers: Vec<Box<dyn Parser<T = Command>>> = vec![];
        if self.can_score(player) {
            parsers.push(Box::new(score_parser(crate::available_scores(
                &self.remaining_dice,
            ))));
        }
        if self.can_roll(player) {
            parsers.push(Box::new(roll_parser()));
        }
        if self.can_done(player) {
            parsers.push(Box::new(done_parser()));
        }
        if parsers.is_empty() {
            None
        } else {
            Some(Box::new(OneOf::new(parsers)))
        }
    }
}

pub fn score_parser(available: Vec<Score>) -> impl Parser<T = Command> {
    let available: Vec<Box<dyn Parser<T = Vec<Die>>>> = available
        .into_iter()
        .map(|s| Box::new(score_dice_parser(s)) as Box<dyn Parser<T = Vec<Die>>>)
        .collect();
    Map::new(
        Chain2::new(
            Doc::name_desc("score", "score dice", Token::new("score")),
            AfterSpace::new(Doc::name_desc(
                "dice",
                "the dice to score",
                OneOf::new(available),
            )),
        ),
        |(_, dice)| Command::Score { dice },
    )
}

pub fn score_dice_parser(score: Score) -> impl Parser<T = Vec<Die>> {
    let name: String = score.dice.iter().map(|d| d.name()).collect();
    let desc = format!("{} points", score.value);
    let token_name = name.clone();
    Map::new(
        Doc::name_desc(name, desc, Token::new(token_name)),
        move |_| score.dice.clone(),
    )
}

pub fn roll_parser() -> impl Parser<T = Command> {
    Map::new(
        Doc::name_desc("roll", "roll the dice", Token::new("roll")),
        |_| Command::Roll,
    )
}

pub fn done_parser() -> impl Parser<T = Command> {
    Map::new(
        Doc::name_desc("done", "finish your turn", Token::new("done")),
        |_| Command::Done,
    )
}
